Filter Kubernetes Nodes on multiple fields

dotNomad opened the pull request.

dotNomad committed:
Expand ListConfig to get/set filters (SOC-9450)

dotNomad committed:
Show and select filters from ListConfig (SOC-9450)

dotNomad committed:
Reduce filters to single data source (SOC-9450)

- Use action to set the filterKey
- if the filterKey is not set use the default filter

dotNomad committed:
Expand IListConfig getFilters comment (SOC-9450)

dotNomad commented:
Added a list of filters to the `IListConfig` interface that contains all of the fields we want to filter on. This creates a selection field that changes what the filter string filters on.
The transformEntities of the `kubernetes-nodes-list-config.service.ts` consumes the `filterKey` (the id of the filter we are using) to decide how to filter the Data Source.

jeremy-moffitt commented:
looks like either something is missing or the unit tests need updated:

> Property 'getFilters' is missing in type '{ allowSelection: false; cardComponent: null; defaultView: ListView; enableTextFilter: false; getColumns: () => any; getDataSource: () => any; getGlobalActions: () => any; getInitialised: () => any; ... 6 more ...; viewType: ListViewTypes.BOTH; }'

dotNomad committed:
Update list component tests (SOC-9450)

dotNomad committed:
Handle no default filter selection better (SOC-9450)

- Disable filter string if no default filter is selected

dotNomad committed:
Only update list filter if a default is found (SOC-9450)

GarySmith reviewed (commented):
We talked about making `getFilters` optional, which would avoid requiring adding the empty method to all of these places. Is that still a possibility?

dotNomad commented:
I did attempt to make the `getFilters` optional, but ran into some other issues that would have forced type definition changes or change implements / extends on classes. Keeping `getFilters` felt like the less impactful change.

richard-cox commented on custom-src/frontend/app/custom/kubernetes/list-types/kubernetes-nodes/kubernetes-nodes-list-config.service.ts:
```diff
@@ -68,28 +76,79 @@ export class KubernetesNodesListConfigService implements IListConfig<KubernetesN
+    const transformEntities: DataFunction<KubernetesNode>[] = [
+      (entities: KubernetesNode[], paginationState: PaginationEntityState) => {
+        const filterKey = paginationState.clientPagination.filter.filterKey;
+        const filterString = paginationState.clientPagination.filter.string.toUpperCase();
```
If `filterString` is a falsy can return `entities` without continuing

richard-cox commented on custom-src/frontend/app/custom/kubernetes/list-types/kubernetes-nodes/kubernetes-nodes-list-config.service.ts:
```diff
@@ -68,28 +76,79 @@ export class KubernetesNodesListConfigService implements IListConfig<KubernetesN
+        switch (filterKey) {
+          case KubernetesNodesListFilterKeys.IP_ADDRESS:
+            return entities.filter(node => {
+              const internalIP: string = node.status.addresses.find(address => {
```
I'm not up to speed with kube spec, but would it be possible to have multiple or no addresses marked at `InternalIP`. If so would be worth covering both cases just in case

dotNomad replied:
It looks like its expected to only be a string, not a list of IPs, and documentation only references a single `InternalIP`
https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.15/#nodeaddress-v1-core

It does look like it could be `ExternalIP` or `InternalIP` though so we could have a separate filter for either and handle it better if one isn't present.

richard-cox replied:
I've pushed a change that will fall back on the `ExternalIP` and also handle the case where neither are found. If the `ExternalIP` check goes against the spec of this feature or you'd rather it not there just let me know and will pull it out.

richard-cox commented on custom-src/frontend/app/custom/kubernetes/list-types/kubernetes-nodes/kubernetes-nodes-list-config.service.ts:
```diff
@@ -68,28 +76,79 @@ export class KubernetesNodesListConfigService implements IListConfig<KubernetesN
+    const transformEntities: DataFunction<KubernetesNode>[] = [
+      (entities: KubernetesNode[], paginationState: PaginationEntityState) => {
+        const filterKey = paginationState.clientPagination.filter.filterKey;
+        const filterString = paginationState.clientPagination.filter.string.toUpperCase();
```
To avoid converting each property/properties in entities to uppercase each time we pass through transform the filterString could be a regex that ignores case. So something like...
`const filterRegEx = new RegExp(filterString, "i")`
`<x>.search(filterRegEx)`
We probably haven't done this in other places when we should have.

dotNomad replied:
Good change to clean up the code, but does add regex support to the filtering which could be confusing if not apparent to the user. I'm initially thinking about IP Addresses where the `.` would the regex dot.
We could escape the filterString so the `.` would actually filter for `.` (and any other special regex characters), but that just adds back in code complexity.

richard-cox replied:
Good point! Lets stick with what's there.

richard-cox commented on src/frontend/packages/store/src/reducers/pagination-reducer/pagination-reducer-set-client-filter-key.ts:
```diff
@@ -0,0 +1,22 @@
+            ...clientPagination,
+            filter: {
+                ...clientPagination.filter,
+                items: {
```
Think the `items` spread can go

richard-cox commented on src/frontend/packages/core/src/shared/components/list/list.component.html:
```diff
@@ -201,4 +215,11 @@
\ No newline at end of file
+</div>
+
+<ng-template #refreshButton>
```
Probably a merge artefact, can be deleted

dotNomad replied:
Good catch.

richard-cox reviewed (changes requested):
I've done a review of the PR, overall looking good! There's some niggly things inline. Additionally....
- We've spoken to Sorin and raised a couple of things to look at
  - The position of the `Filter Selection` text
  - General pattern of filtering by data that's not shown in the table (ip address & labels)
- ~At some point it would be nice to port this upstream, so limiting the changes such as getFilters would be good. I'll look into this and see if there's a clean way of making it optional~
- ~I need to come back and test some more edge case like scenarios~
- ~It looks like the pagination e2e tests are failing due to the additional mat select in the dom~
  - see `application-wall-e2e.spec.ts` createPageSizeSelectId
  - The `mat-select-<x>` will probably need to be tweaked, a better solution would be better selectors but for now lets low bar
  - I'll send Instructions on how to run e2e tests in RC

dotNomad committed:
Remove unnecessary additions (SOC-9450)

dotNomad committed:
Return entities if no filter string (SOC-9450)

dotNomad committed:
Adjust test mat-select number

Adding an additional mat-select created an off-by-one issue

dotNomad committed:
Remove unintended fdescribe

richard-cox committed:
Fall back on `ExternalIP` if `InternalIP` is not found, make result null safe

richard-cox committed:
Make getFilters optional
- make optional in interface
- check for optional function in list component, even though in class it's not optional
- fixed unrelated typing for getMultiActions and getSingleActions

richard-cox commented:
We've had a round of review with Sorin and would like to address the following points
- Present information in the table that is being searched on
  - Add `IP` column with globe icon and ip address in pop over
  - Add `Labels` column containing an `app-chips` component. This will show chips that fit into the space with an option to show all
  - Combine `Disk Pressure` and `Memory Pressure` columns into one
  - See examples in https://xd.adobe.com/spec/dcd4203c-d334-462e-7921-ea6c3ef41151-aae5/screen/a2ea0108-1ae5-418f-8153-bc2a1fe985d5/Filter/ for details
- Retain the `Filter selection` for now and look for user input

richard-cox committed:
Merge remote-tracking branch 'origin/master' into SOC-9450

# Conflicts:
#	custom-src/frontend/app/custom/kubernetes/list-types/kubernetes-nodes/kubernetes-nodes-data-source.ts
#	src/frontend/packages/cloud-foundry/src/shared/components/list/list-types/cf-services/cf-services-list-config.service.ts
#	src/frontend/packages/cloud-foundry/src/shared/components/list/list-types/cf-users-org-space-roles/cf-users-space-roles-list-config.service.ts
#	src/frontend/packages/cloud-foundry/src/shared/components/list/list-types/detach-apps/detach-apps-list-config.service.ts
#	src/frontend/packages/core/src/shared/components/list/list-types/github-commits/github-commits-list-config-base.service.ts
#	src/frontend/packages/core/src/shared/components/list/list.component.ts
#	src/frontend/packages/store/src/actions/pagination.actions.ts

richard-cox committed:
Update following design doc
- https://xd.adobe.com/spec/dcd4203c-d334-462e-7921-ea6c3ef41151-aae5/screen/a2ea0108-1ae5-418f-8153-bc2a1fe985d5/Filter/
- Include optional labels list component
- Fix issue where search term was not capitalised
- Improve typing

richard-cox committed:
E2E Fix
